examples/profiling/primitive_instrumentation.cpp

```diff
@@ -0,0 +1,205 @@
+            define(error, constant(0.0, shape(x, 0))),
+            define(gradient, constant(0.0, shape(x, 1))),
+            define(step, 0),
+            while(
+                step < 10,
+                block(
+                    store(pred, 1.0 / (1.0 + exp(-dot(x, weights)))),
+                    store(error, pred - y),                  // error: [30]
+                    store(gradient, dot(transx, error)),     // gradient: [2]
+                    parallel_block(
+                        store(weights, weights - (alpha * gradient)),
+                        store(step, step + 1)
+                    )
+                )
+            ),
+            weights
+        )
+    ),
+    lra
+))";
+
+// Retrieve performance counter data for selected primitives
+//   primitive_instances:     The primitives for which performance counter data
+//                            is required
+//   locality_id:             The locality the performance counter data is going
+//                            to be queried from
+//   counter_name_last_parts: A vector containing the last part of the
+//                            performance counter names.
+//                            e.g. std::vector{ "count/eval", "count/direct_eval" }
+// NOTE: primitive_instances are not verified
+std::map<std::string, std::vector<std::int64_t>> retrieve_counter_data(
+    std::vector<std::string> const& primitive_instances,
+    hpx::naming::id_type locality_id,
+    std::vector<std::string>
+        counter_name_last_parts)
+{
+    // Return value
+    std::map<std::string, std::vector<std::int64_t>> result;
+
+    // Reuse get_counter_values_array calls
+    //   key: primitive type
+    //   value: vector of performance counter values
+    std::map<std::string, std::vector<std::vector<std::int64_t>>>
+        counter_values_pile;
+
+    // Iterate through all provided primitive instances
+    for (auto const& name : primitive_instances)
+    {
+        // Parse the primitive name
+        auto tags =
+            phylanx::execution_tree::compiler::parse_primitive_name(name);
+
+        // TODO: Ensure counter_name_last_part has at least one entry
+
+        // Performance counter values
+        std::vector<std::vector<std::int64_t>>& counter_values =
+            counter_values_pile[tags.primitive];
+        if (counter_values.empty())
+        {
+            // Iterate through the last parts of performance counter names
+            for (auto const& counter_name_last_part : counter_name_last_parts)
+            {
+                // Construct the name of the counter
+                std::string counter_name("/phylanx/primitives/" +
+                    tags.primitive + "/" + counter_name_last_part);
+                // The actual performance counter
+                hpx::performance_counters::performance_counter counter(
+                    counter_name, locality_id);
+                counter_values.push_back(
```

Comment:
You could request the counter values asynchronously:
```
        if (counter_values.empty())
        {
            // Iterate through the last parts of performance counter names
            std::vector<hpx::future<std::vector<counter_array_value>>> futures;
            futures.reserve(counter_name_last_parts.size());
            for (auto const& counter_name_last_part : counter_name_last_parts)
            {
                // Construct the name of the counter
                std::string counter_name("/phylanx/primitives/" +
                    tags.primitive + "/" + counter_name_last_part);
                // The actual performance counter
                hpx::performance_counters::performance_counter counter(
                    counter_name, locality_id);
                futures.push_back(counter.get_counter_values_array(false));
            }

            hpx::wait_all(futures());
            for (auto& f : futures)
            {
                counter_values.push_back(f.get().values));
            }
       }
```

Reply:
You could even rewrite things such that you request all counters (for all primitives) first before calling `wait_all()`.

Reply:
Done.